packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -578,6 +641,99 @@ export class TokenDetectionController extends StaticIntervalPollingController<To
     return slicesOfTokensToDetect;
   }
 
+  /**
+   * This adds detected tokens from the Accounts API, avoiding the multi-call RPC calls for balances
+   * @param options - method arguments
+   * @param options.selectedAddress - address to check against
+   * @param options.chainId - chainId to check tokens for
+   * @param options.tokenCandidateSlices - these are tokens we know a user does not have (by checking the tokens controller).
+   * We will use these these token candidates to determine if a token found from the API is valid to be added on the users wallet.
+   * It will also prevent us to adding tokens a user already has
+   * @returns a success or failed object
+   */
+  async #addDetectedTokensViaAPI({
+    selectedAddress,
+    chainId,
+    tokenCandidateSlices,
+  }: {
+    selectedAddress: string;
+    chainId: Hex;
+    tokenCandidateSlices: string[][];
+  }) {
+    return await safelyExecute(async () => {
+      const tokenBalances = await this.#accountsAPI
+        .getMultiChainBalances(selectedAddress, chainId)
+        .catch(() => null);
+
+      if (!tokenBalances || tokenBalances.length === 0) {
+        return { result: 'failed' } as const;
+      }
+
+      const tokensWithBalance: Token[] = [];
+      const eventTokensDetails: string[] = [];
+
+      const tokenCandidateSet = new Set<string>(tokenCandidateSlices.flat());
+
+      tokenBalances.forEach((token) => {
+        const tokenAddress = token.address;
+
+        // Make sure that the token to add is in our candidate list
+        // Ensures we don't add tokens we already own
+        if (!tokenCandidateSet.has(token.address)) {
+          return;
+        }
+
+        // We need specific data from tokenList to correctly create a token
+        // So even if we have a token that was detected correctly by the API, if its missing data we cannot safely add it.
+        if (!this.#tokenList[token.address]) {
+          return;
+        }
+
+        // NOTE: some of this data is also available from the `token` balance,
+        // however it is missing the `aggregators` and `iconUrl` properties
+        const { decimals, symbol, aggregators, iconUrl, name } =
+          this.#tokenList[token.address];
+        eventTokensDetails.push(`${symbol} - ${tokenAddress}`);
+        tokensWithBalance.push({
+          address: tokenAddress,
+          decimals,
+          symbol,
+          aggregators,
+          image: iconUrl,
+          isERC721: false,
+          name,
+        });
+      });
+
+      if (tokensWithBalance.length) {
+        this.#trackMetaMetricsEvent({
```

Comment:
To DRY or not to DRY.

Meh, this is only repeated for a second time. Opting to keep as is.